sound/soc/sof/topology.c

```diff
@@ -1303,6 +1351,27 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 		return ret;
 	}
 
+	/* if there is only one source/sink pin, there is no need to parse source/sink
+	 * pin binding from topology.
+	 */
+	if (swidget->num_source_pins > 1) {
+		ret = sof_parse_pin_binding(swidget, priv, true);
+		if (ret < 0) {
+			dev_err(scomp->dev, "failed to parse component pin tokens for %s\n",
+				w->name);
+			return ret;
+		}
+	}
+
+	if (swidget->num_sink_pins > 1) {
+		ret = sof_parse_pin_binding(swidget, priv, false);
+		if (ret < 0) {
+			dev_err(scomp->dev, "failed to parse component pin tokens for %s\n",
+				w->name);
+			return ret;
+		}
+	}
+
```

Comment:
if there's an error in any part of the code below, you'd need to free both src/sink binding array. maybe do this as the last step in widget_ready so you dont have to deal with all the frees

Reply:
I am wondering why asoc-topology doesn't help to call widget_free, and free the widget totally.